Lib/pathlib/_types.py

```diff
@@ -0,0 +1,54 @@
+"""
+Protocols for supporting classes in pathlib.
+"""
+from typing import Protocol, runtime_checkable
+
+
+@runtime_checkable
+class Parser(Protocol):
+    """Protocol for path parsers, which do low-level path manipulation.
+
+    Path parsers provide a subset of the os.path API, specifically those
+    functions needed to provide PurePathBase functionality. Each PurePathBase
+    subclass references its path parser via a 'parser' class attribute.
+    """
+
+    sep: str
+    def join(self, path: str, *paths: str) -> str: ...
+    def split(self, path: str) -> (str, str): ...
+    def splitdrive(self, path: str) -> (str, str): ...
+    def splitext(self, path: str) -> (str, str): ...
+    def normcase(self, path: str) -> str: ...
+    def isabs(self, path: str) -> bool: ...
+
+
+@runtime_checkable
+class DirEntry(Protocol):
+    """Protocol for directory entries, which store information about directory
+    children.
+
+    Directory entries provide a subset of the os.DirEntry API, specifically
+    those methods and attributes needed to provide PathBase functionality
+    like glob() and walk(). Directory entry objects are generated by
+    PathBase.scandir().
+
+    Path and PathBase implement this protocol, but their is_*() methods always
+    fetch up-to-date information.
+    """
+
+    name: str
+    def is_dir(self, *, follow_symlinks: bool = True) -> bool: ...
```

Comment:
Shouldn't this include `is_file()` as well? In case you don't want to touch special file types? (nor treat deleted directory as a file). Adding this later wouldn't be backwards compatible.

Reply:
I'm OK with this. Should we include `is_junction()` too perhaps?

Reply:
No, most implementations would need to always return True. But maybe we can generalise symlinks and junctions, if you don't care about the difference and just don't want to follow any type of link.